### DISCUSSION TITLE
Advice on supervised ID tracking

### DISCUSSION THREAD

**Discussion by prashers** [talmolab/sleap#1149] [Help!]
Hi SLEAP team,

I am attempting to use SLEAP to track individuals within groups of six quails in a flight pen. Each bird was wearing a unique lettered backpack that is visible in most frames (although the letter is obscured when the birds cluster or when there is glare due to the backpacks being laminated). 

So far, @talmo has guided me through training supervised ID models and getting predictions from them. I have gone through a few rounds of training and correcting predictions and now have about 400 labelled frames. However, the accuracy of predictions is not as good as I had hoped - not all birds in a frame get a predicted label and their assigned tracks are not always correct (see screenshot below). 

![image](https://user-images.githubusercontent.com/65912929/214380105-0fcacb31-66b6-4331-ae73-3a32266eeb80.png)

So I was wondering:

1. Is there a general estimate of how many frames I should label for accurate predictions on individual frames (or does this really depend on my particular situation)?

2. I have been labeling birds even if their lettered tag is not visible in a particular frame (but I still know which bird it is). If I want SLEAP to identify birds and assign them to the correct track based on their backpack IDs, should I only label birds when their tags are visible (I worry that I am introducing too much noise in the training data if I label and assign birds to tracks when their tags are not visible)?

Thank you very much! Let me know if you need more information.

**Comment by talmo** [comment]
Hi @prashers,

The ID models are a bit unstable still unfortunately, so there's a lot of trial and error with the hyperparameters. I'd recommend experimenting with the loss weights through the config JSON file.

Are you using top-down? If so, do you mind attaching the last `training_config.json` for the centered instance model?

In terms of your questions:

1. Number of frames totally depends on your situation, but keep in mind the best practices for labeling, namely that it's essential to maximize diversity and cover the cases where the model isn't doing well instead of frames that are easy to label.
2. Labeling ones that have the backpack occluded should be ok. Naturally the accuracy might not be as good, but it's worth it to include it so the model sees some ambiguous examples.

Let us know about the configs you're rocking at the moment and we can try to tweak from there!

Cheers,

Talmo

**Comment by prashers** [comment]
Yes, I am using the top-down approach. I've attached the training_config.json file.

Thanks for the info!
________________________________

**Comment by talmo** [comment]
Hi @prashers,

Thanks for sending that over. A couple of things:

1. Since this is a top-down model, if the birds aren't detected then that's an issue with the centroid model. You can try tweaking that architecture to improve things. Try setting the sigma to 3.5 or 5 for the centroid model, increasing the max stride to 32 and reducing the filter rate to 1.5 and retraining.
2. For the ID being incorrect: this will happen occasionally, but you can certainly try to push on the accuracy a bit more. In the JSON file for the centered instance ID model, try increasing the loss weight to 0.01:
![image](https://user-images.githubusercontent.com/3187454/214908010-634d312e-c7f3-4a25-848d-f774fc7f6b31.png)
This might decrease the pose performance, but it's worth giving it a go and seeing how it does with ID.

One more thing: for both models, set the rotation augmentation to -180 to 180, which in your case should help with generalizing to new data a bit better.

Let us know how it goes after trying these!

Cheers,

Talmo

**Reply by prashers** [reply]
Those changes definitely helped(!), but I still see some of the same types of errors. I think I do need to focus on labeling the troublesome frames like you said. 

I have one final question for now - for the videos I already have, I had been reusing two sets of lettered tags for three groups each (each group contained uniques individuals). I would definitely need to train different models for the two sets of letters. However, would I be able to use the same models across groups that had the same lettered tags if I make sure to include labelled frames for each group when training the models? I hope that makes sense.

Thanks,
Sanjay

**Reply by roomrys** [reply]
Hi @prashers,

To clarify, you want to train the model using examples from one group and use this trained model to generate predictions on the new group (where both groups use the same set of lettered tags). This should be doable. I am curious to see how much of the animal's identity is weighted by the unique marker alone (especially since it is used as the location of a node in your skeleton).

>Note: The same rules about video similarity apply here: to have similar results from a trained mode, we need the videos to be similar - more examples/labels across videos will help generalize the model. 

Any input @talmo?

Thanks,
Liezl

**Reply by prashers** [reply]
Hi @talmo and @roomrys ,

I took a break from this for a while, but was recently trying to train a supervised ID model for new videos I have in which my birds are wearing larger backpacks with ArUco markers on them. 

I was initially trying to use NAPS (https://github.com/kocherlab/naps) in conjunction with SLEAP to assign tracks based on the ArUco markers, but an issue with that is that the markers are not always detected accurately (esp. when the birds are clustering or a feather partly overlaps the marker). I decided to try the top-down supervised ID model again, hoping it could handle these situations better. 

With the supervised ID models, I am still facing the issue of getting frames with no instances on some birds. I have tried adjusting the centroid model parameters according to Talmo's suggestions above. I have tried a max stride of 16 and 32, a filters rate as low as 1.00, and a sigma of 3.5 and 5. 
I also have rotation augmentation set to -180 to 180, contrast from 0.5 to 2, and brightness from 0 to 10.

Are there any other values I could try with these parameters, or other parameters I could play with?

Thanks!

**Comment by roomrys** [comment]
Hi @prashers,

In your original image, I noticed that the frame was unable to pick-up the instance with the "K" marker, but we see that the instances tab has another instance labeled as "track_K". My hypothesis (not yet proven) is that SLEAP decided the other instance marked as "track_K" had a higher likelihood of being track_K (although not tagged with the K marker) which meant that the instance which is tagged with "K" could not be assigned to it's true track "track_K" and was perhaps thrown out as undetected.

https://github.com/talmolab/sleap/blob/5ba6bc12b3663eb947fc1e5d2adb6b4cf9c4481a/sleap/nn/identity.py#L60-L80

What is the other tag? You have "A", "B", "F", "K", "S" and what else? It isn't actually another "K" is it? 

When you notice these undetected instances, is it always the case that one of the tracks is assigned incorrectly? If so, we will need to take a deeper look at the ID tracking as we most definitely do not want to throw out detections. For your own dataset, it points to mislabeling in the tracks of instances - unless the unknown tag looks very similar to a "K".

Thanks,
Liezl

**Reply by prashers** [reply]
In those older videos, I was using tags A, B, C, F, K and S. 
On that particular frame, the bird with tag C was incorrectly assigned track_K.

No, it is not always the case that one track is assigned incorrectly.

Below are a couple examples from my newer videos. 

The first is another case where a track was assigned incorrectly (ArUcoTag#1.0 should be assigned to the middle bird that is missing an instance).
In the second frame below, the tracks were correctly assigned for 5 birds, but the 6th bird did not get an instance and track (here, the 6th bird is ArUcoTag#1.0 again, but it is not always the same bird that is missing an instance/track). The nodes on a couple instances are not very accurate in this one either.

Thanks,
Sanjay

![image](https://github.com/talmolab/sleap/assets/65912929/1c869e10-8d98-451b-b925-01d1b44d3213)


![image](https://github.com/talmolab/sleap/assets/65912929/0ec8ddd7-889c-4edb-b294-c0cb5b569317)
